lib/filewatch/since_db.rb

```diff
@@ -0,0 +1,247 @@
+
+    attr_reader :path, :short_keys
+    attr_writer :logger
+
+    def initialize(opts, loggr)
+      @logger = loggr
+      @path = opts[:sincedb_path]
+      @lastwarn = Hash.new { |h, k| h[k] = 0 }
+      @sincedb_last_write = 0
+      @interval = opts[:sincedb_write_interval]
+      @sincedb = {}
+      @short_keys = Hash.new{|h, k| h[k] = []}
+    end
+
+    def write_periodically
+      now = Time.now.to_i
+      delta = now - @sincedb_last_write
+      if delta >= @interval
+        @logger.debug? && @logger.debug("writing sincedb (delta since last write = #{delta})")
+        sincedb_write
+        @sincedb_last_write = now
+      end
+    end
+
+    def write(reason=nil)
+      @logger.debug? && @logger.debug("caller requested sincedb write (#{reason})")
+      sincedb_write
+    end
+
+    def open
+      @time_sdb_opened = Time.now.to_f
+      begin
+        File.open(path) do |db|
+          @logger.debug? && @logger.debug("SinceDb open: reading from #{path}")
+          db.each {|line| store_from_file(line) }
+        end
+
+        @logger.debug? && @logger.debug("SinceDb open: keys read #{@sincedb.keys.inspect}")
+      rescue => e
+        #No existing sincedb to load
+        @logger.debug? && @logger.debug("SinceDb open: error: #{path}: #{e.inspect}")
+      end
+    end
+
+    def deallocate(wf)
+      return if !member?(wf.storage_key)
+      get(wf.storage_key).deallocate_watched_file
+    end
+
+    def find(wf, event)
+      @logger.debug? && @logger.debug("SinceDb find: event: #{event}, path: #{wf.path}")
+      key = wf.storage_key
+      value = get(key)
+      if value
+        @logger.debug? && @logger.debug("SinceDb find: found on first fp - #{key} => #{value}")
+        # does the on disk record have a second fp?
+        # then check it
+        fp2 = value.second_fingerprint
+        if fp2
+          if !wf.last_fingerprint_match?(fp2)
+            value = nil
+          else
+            @logger.debug? && @logger.debug("SinceDb find: matched on second fp:#{fp2}")
+          end
+        end
+      end
+      if value
+        if value.watched_file.nil?
+          # unallocated as read from disk
+          @logger.debug? && @logger.debug("SinceDb find: allocating - #{value}")
+          wf.update_bytes_read(value.position)
+          value.upd_watched_file(wf)
+        elsif value.watched_file == wf
+          @logger.debug? && @logger.debug("SinceDb find: exact match - #{value}")
+          # or allocated to this watched_file via the SinceDbConnverter v1 conversion
+          value.upd_expiry
+        else
+          @logger.debug? && @logger.debug("SinceDb find: match but allocated to another - #{value}")
+          # we found value having the same key as wf
+          # wf is a renamed file recently discovered
+          # or a different file with the same content
+          # leave as is
+        end
+        @logger.debug? && @logger.debug("SinceDb find: returning value")
+        return value
+      end
```

Comment:
I am a bit confused by the flow of this method and the check against `value` Could we split up the logic in small methods?